### PR TITLE
apply comma split logic to servlet request extract adapter

### PR DESCRIPTION
**Pull request by fujigon**

### CODE REVIEW

**Comment by fujigon** [comment]
In some cases, multiple values for the same key in http header as multiple lines (A) or comma separated list (B).
cf. https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
In case of B, tomcat servlet returns comma separated value like `xxx, yyy` for the call ``carrier.getHeader`, and leads to failure while parsing the value at https://github.com/DataDog/dd-trace-java/blob/master/dd-trace-ot/src/main/java/datadog/opentracing/propagation/B3HttpCodec.java#L104

**Comment by devinsba** [comment]
Do you know what the source of the multiple instances of `X-B3-Sampled` is? That seems like a bug that needs to be addressed also.

**Comment by fujigon** [comment]
> Do you know what the source of the multiple instances of X-B3-Sampled is? That seems like a bug that needs to be addressed also.

I'm not for sure, but I feel that the two headers are injected by APM's instrument for apache-http-client and APM's instrument for jaxrs-client(jersey-client).
We use [jaxrs-client](https://github.com/eclipse-ee4j/jersey/blob/master/core-client/src/main/java/org/glassfish/jersey/client/JerseyClient.java), and internal of the jersey-client, employ [apache connector](https://github.com/eclipse-ee4j/jersey/tree/master/connectors/apache-connector/src/main/java/org/glassfish/jersey/apache/connector).

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/TomcatServlet3Test.groovy [hunk 1]
```diff
@@ -297,7 +297,7 @@ abstract class TomcatDispatchTest extends TomcatServlet3Test {
           tags {
             "$Tags.COMPONENT" serverDecorator.component()
             "$Tags.SPAN_KIND" Tags.SPAN_KIND_SERVER
-            "$Tags.PEER_HOSTNAME" "localhost"
+            "$Tags.PEER_HOSTNAME" { it == "localhost" || it == "127.0.0.1" }
```
Comment (fujigon) [inline dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/TomcatServlet3Test.groovy]: https://github.com/DataDog/dd-trace-java/blob/57c3ec196f6a56545ed478d656f92f14fef8dc7a/dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/JettyServlet3Test.groovy#L249

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy [hunk 2]
```diff
@@ -232,6 +232,42 @@ abstract class HttpServerTest<SERVER, DECORATOR extends HttpServerDecorator> ext
     body = null
   }
 
+  def "test success with multiple header attached parent"() {
```
Comment (fujigon) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy]: parent with multiple headers (comma separated values).

Comment (fujigon) [reply dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/base/HttpServerTest.groovy]: confirmed locally to run `TomcatServlet3Test.groovy`  and 
- this test case passes with [my change](https://github.com/DataDog/dd-trace-java/pull/1131/files#diff-182f6c96b9852155296cb24a4f269bfa)
- this test case fails without my change.

**Comment by fujigon** [comment]
about https://github.com/DataDog/dd-trace-java/pull/1131#issuecomment-563654258

https://github.com/DataDog/dd-trace-java/blob/42b88c7b8ec99b8b9ac2a4d4dda90ebbbe9e5921/dd-java-agent/instrumentation/apache-httpclient-4/src/main/java/datadog/trace/instrumentation/apachehttpclient/HttpHeadersInjectAdapter.java#L12

might be the cause.

**Comment by fujigon** [comment]
@devinsba @tylerbenson 
Hi, I added the test and updated the change.
Could you kindly have a look?

**Comment by devinsba** [comment]
One of the failing tests is a known failure. The two others appear to have occurred due to adding another test to the http server suite. I think having the test is important so I wouldn't remove it

@DataDog/apm-java thoughts?

**Comment by tylerbenson** [comment]
I just thought of a case where this might change behavior for someone... if they are using `dd.trace.header.tags`, this will result in a possible fragment of the original tag being added.

@DataDog/apm-java Think this is something we should be concerned about?

**Comment by randomanderson** [comment]
> I just thought of a case where this might change behavior for someone... if they are using `dd.trace.header.tags`, this will result in a possible fragment of the original tag being added.
> 
> @DataDog/apm-java Think this is something we should be concerned about? In my opinion, it seems like it would be better to break that and fix this.

I agree.  I would go further and say that anyone relying on a comma separated header is asking for trouble because the comma denotes multiple headers of the same name.  For example:

```
X-Foo: bar, bas
```

can also be

```
X-Foo: bar
X-Foo: bas
```

We don't specify which of the 2 will become the tag.

I do wish we new why multiple injections were happening though.

**Comment by fujigon** [comment]
@tylerbenson 
Thanks for the update! I rebased onto current master. :)

**Pull request merged**